bigtable/client/metadata_update_policy.h

```diff
@@ -0,0 +1,106 @@
+// Copyright 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_METADATA_UPDATE_POLICY_H_
+#define GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_METADATA_UPDATE_POLICY_H_
+
+#include <bigtable/client/version.h>
+
+#include <grpc++/grpc++.h>
+#include <memory>
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+/**
+ * Define the class for governing x-goog-request-params metadata value.
+ *
+ * The value of x-goog-request-params starts with one of the following suffix
+ *    "parent=" : Operation in instance, e.g. TableAdmin::CreateTable.
+ *    "table_name=" : table_id is known at the time of creation, e.g.
+ *     Table::Apply.
+ *    "name=" : this is used when table|_id is known only in the RPC call, e.g.
+ *     TableAdmin::GetTable.
+ *
+ */
+class MetadataParamTypes final {
+ public:
+  static MetadataParamTypes const PARENT;
+  static MetadataParamTypes const NAME;
+  static MetadataParamTypes const TABLE_NAME;
+
+  bool operator==(MetadataParamTypes const& that) const {
+    return type_ == that.type_;
+  }
+  std::string const& getType() const { return type_; }
+
+ private:
+  std::string type_;
+  MetadataParamTypes(std::string type) : type_(std::move(type)) {}
+};
+
+/// MetadataUpdatePolicy holds supported metadata and setup ClientContext
+class MetadataUpdatePolicy {
+ public:
+  /**
+   * Constructor with default metadata pair.
+   *
+   * @param resource_name hierarchical name of resource, including  project id,
+   * instance id
+   *        and/or table_id.
+   * @param metadata_param_type type to decide prefix for the value of
+   *     x-goog-request-params
+   */
+  MetadataUpdatePolicy(std::string resource_name,
+                       MetadataParamTypes metadata_param_type);
+
+  /**
+   * Constructor with default metadata pair.
+   *
+   * @param resource_name hierarchical name of resource, including  project id,
+   * instance id
+   *        and/or table_id.
+   * @param metadata_param_type type to decide prefix for the value of
+   *     x-goog-request-params.
+   * @param table_id table_id used in RPC call.
+   */
+  MetadataUpdatePolicy(std::string resource_name,
+                       MetadataParamTypes metadata_param_type,
+                       std::string table_id);
+
+  MetadataUpdatePolicy(MetadataUpdatePolicy&& rhs) noexcept = default;
```

Comment:
I believe you do not need to define these, the compiler provides the implementation.

Reply:
i am getting following exception if it is not defined.

`/home/manish/git/google-cloud-cpp/bigtable/client/table.cc: In member function ‘bigtable::v0::RowReader bigtable::v0::Table::ReadRows(bigtable::v0::RowSet, bigtable::v0::Filter)’:
/home/manish/git/google-cloud-cpp/bigtable/client/table.cc:144:68: error: use of deleted function ‘bigtable::v0::RowReader::RowReader(bigtable::v0::RowReader&&)’
                        bigtable::internal::ReadRowsParserFactory>());`

Reply:
Ack.  Leave them there then.  But you have another build problem (e.g. [here](https://travis-ci.org/GoogleCloudPlatform/google-cloud-cpp/jobs/354164706#L3747)), you may need to remove the `noexcept` exception specification from `operator=`.

Reply:
fixed